codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.http2;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+import io.netty.channel.AbstractChannel;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelConfig;
+import io.netty.channel.ChannelMetadata;
+import io.netty.channel.ChannelOutboundBuffer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.DefaultChannelConfig;
+import io.netty.channel.EventLoop;
+import io.netty.channel.RecvByteBufAllocator;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.internal.InternalThreadLocalMap;
+import io.netty.util.internal.OneTimeTask;
+
+import java.net.SocketAddress;
+import java.nio.channels.ClosedChannelException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * Child {@link Channel} of another channel, for use for modeling streams as channels.
+ */
+abstract class AbstractHttp2StreamChannel extends AbstractChannel {
+    /**
+     * Used by subclasses to queue a close channel within the read queue. When read, it will close
+     * the channel (using Unsafe) instead of notifying handlers of the message with {@code
+     * channelRead()}. Additional inbound messages must not arrive after this one.
+     */
+    protected static final Object CLOSE_MESSAGE = new Object();
+    private static final ChannelMetadata METADATA = new ChannelMetadata(false);
```

Comment:
@ejona86 - In summary the `RecvByteBufAllocator` was previously limited to enforcing the number of reads per event loop read, but its abilities were made more general by https://github.com/netty/netty/commit/cf171ff52555b9e984a3b9103287f6b897dc8626.  Previously there was an [instanceof in DefaultChannelConfig](https://github.com/netty/netty/commit/cf171ff52555b9e984a3b9103287f6b897dc8626#diff-2698a3e4d66a69f17d6179c7fc467fd0L71) to determine the minimum amount of reads based on the type of channel.  Instead of having `DefaultChannelConfig` know details about the channel types this was moved into [ChannelMetaData](https://github.com/netty/netty/commit/cf171ff52555b9e984a3b9103287f6b897dc8626#diff-f88dc4666216aaf16a9243749cccf02bR48). The naming `minMax...` is confusing/unfortunate but is designed to enforce the minimum amount that the maximum must be. The goal is to move channel specific details to be defined in the channel implementations.

> Currently, setRecvByteBufAllocator(RecvByteBufAllocator) is part of ChannelConfig but setRecvByteBufAllocator(RecvByteBufAllocator, ChannelMetadata) is not.

I can add `setRecvByteBufAllocator(RecvByteBufAllocator, ChannelMetadata)` to `ChannelConfig` if that makes things easier. However do you have any alternative suggestions in mind?